### PR TITLE
added implementation for NibbleSlice

### PR DESCRIPTION
**Pull request by nuel-ikwuoma**
PR for NibbleSlice implementation

### CODE REVIEW

File: lib/mpt/NibbleSlice.sol [hunk 1]
```diff
@@ -20,18 +24,75 @@ library NibbleSliceOps {
         return len(self) == 0;
     }
 
-    function eq(NibbleSlice memory self, NibbleSlice memory other) public pure returns (bool) {
+    function eq(NibbleSlice calldata self, NibbleSlice calldata other) public pure returns (bool) {
```
Comment (seunlanlege) [inline lib/mpt/NibbleSlice.sol:27]: Why calldata though?

Comment (nuel-ikwuoma) [reply lib/mpt/NibbleSlice.sol:27]: calldata allows to obtain bytes slices, this is required for the common prefix computation.

File: lib/mpt/NibbleSlice.sol [hunk 2]
```diff
@@ -20,18 +24,75 @@ library NibbleSliceOps {
...
+
+            if(self_align != 0) {
+                if((self.data[self_start] & 0x0F) != (other.data[other_start] & 0x0F)) {
+                    return 0;
+                }
+                ++self_start;
+                ++other_start;
+                ++first;
+            }
+            return biggestDepth(self.data[self_start:], other.data[other_start:]) + first;
```
Comment (nuel-ikwuoma) [inline lib/mpt/NibbleSlice.sol:59]: biggestDepth takes a bytes slice.

File: lib/mpt/NibbleSlice.sol (hunk 1 above)
Comment (seunlanlege) [reply lib/mpt/NibbleSlice.sol:27]: Unfortunately the keys won't be calldata, they'll be constructed in memory. Let's just copy the arrays.

Comment (nuel-ikwuoma) [reply lib/mpt/NibbleSlice.sol:27]: No, not possibe. I'd implement a workaround with inline assembly, as solidity only support bytes slices for calldata types as of now

File: lib/mpt/NibbleSlice.sol (hunk 2 above)
Comment (nuel-ikwuoma) [reply lib/mpt/NibbleSlice.sol]: Fixed slice for memory bytes

**Review by seunlanlege — APPROVED** [review]

**Pull request merged**